### DISCUSSION TITLE
App Logs logcat listener

### DISCUSSION THREAD

**Discussion by john-peterson** [AndroidIDEOfficial/AndroidIDE#1876] [Q&A] [answered]
How does it work?

Sometimes App Logs is empty nothing is picked up

How is it able to listen to another app without READ_LOGS permission

**Comment by itsaky** [comment, marked as answer]
> Sometimes App Logs is empty nothing is picked up

Can't say what actually causes this issue without more information, but you can try the following things :
- Make sure that you install the `debug` variant of your application that has been built with AndroidIDE.
- Enable USB debugging.
- Increase the log buffer size in Android's developer options.

If it still doens't work, file a bug report with more details about the issue.

> How is it able to listen to another app without READ_LOGS permission

This is achieved using AndroidIDE's `logsender` module.

When you build an application in AndroidIDE, the [`LogSender`](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/fac8ac61e16914ce8b60308f0827b2ff2b67fd77/tooling/plugin/src/main/java/com/itsaky/androidide/gradle/LogSenderPlugin.kt#L37) Gradle plugin is applied to the Android application modules (`com.android.application`). This plugin adds the logsender dependency to the debuggable variants of your project ([source](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/fac8ac61e16914ce8b60308f0827b2ff2b67fd77/tooling/plugin/src/main/java/com/itsaky/androidide/gradle/LogSenderPlugin.kt#L83)). This means that _every_ debuggable variant of your _application_ module(s) (not libraries) in your whole project has this dependency.

`ContentProvider`s in Android are started automatically when you open an application. So, the logsender dependency has a [non-exported](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/fac8ac61e16914ce8b60308f0827b2ff2b67fd77/logging/logsender/src/main/AndroidManifest.xml#L33) (private) content provider ([`LogSenderInstaller`](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/fac8ac61e16914ce8b60308f0827b2ff2b67fd77/logging/logsender/src/main/java/com/itsaky/androidide/logsender/utils/LogSenderInstaller.java)), which, when started, spawns a new [background service](https://github.com/AndroidIDEOfficial/AndroidIDE/blob/dev/logging/logsender/src/main/java/com/itsaky/androidide/logsender/LogSenderService.java) to read the logs of your application. When this background service starts (in your application), it _binds_ with the log receiver service in AndroidIDE. When the communication has been established, your application (the service) reads every line from `logcat` and sends these lines to AndroidIDE's log receiver over a local socket connection which are then shown in AndroidIDE's UI.

This works because the `READ_LOGS` permissions is not needed for an application to read its own logs.

Here is a flowchart of the whole process (view on web):

```mermaid
flowchart TD
    A[Build Application in AndroidIDE] --> B[Launch Application]
    B --> C[Start non-exported ContentProvider: LogSenderInstaller]
    C --> D[Spawn Background Service: LogSenderService]
    D --> E[Bind LogSenderService with AndroidIDE's LogReceiverService]
    E --> F[Establish Socket Communication between Application and AndroidIDE]
    F --> G[Read logcat lines in Application Service]
    G --> H[Send logcat lines to AndroidIDE over Local Socket]
    H --> I[Display logs in AndroidIDE's UI]

    subgraph Notes
        note1[READ_LOGS permission is not needed for an application to read its own logs.]
    end

    I -.-> note1

```

I hope this helps.
